### PR TITLE
[RDY] vim-patch:8.1.0614,8.1.0632,8.1.0644,8.1.0658,8.1.0660,8.1.0669,8.1.0673,8.1.0679,8.1.0697,8.1.0701,8.1.0702,8.1.0709,8.1.0717,8.1.0750,8.1.0767,8.1.0772,8.1.0039

### LINKED ISSUE #3803: Regression: screen not correctly redrawn when unplacing sign
Screen is not correctly redrawn when unplacing a sign. 

Steps to reproduce:
```
$ cat >bug.vim <<EOF
set number
call setline(1, ['a', 'b', 'c'])
sign define piet text=>> texthl=Search
exe ":sign place 2 line=2 name=piet file=" . expand("%:p")
EOF
```
```
$ rm /tmp/foo.txt; vim --clean -S bug.vim /tmp/foo.txt
```
Then in vim, type Ex command:
```
:sign unplace 2
```
And observe that screen is not properly drawn.
Pressing CTRL-L works around it.

Bug is a regression introduced by recent patch:
```
commit 27a472c32ed5b5298bca50864570a4a71ec1d204
Author: Bram Moolenaar <Bram@vim.org>
Date:   Wed Jan 9 21:47:30 2019 +0100

    patch 8.1.0709: windows are updated for every added/deleted sign
    
    Problem:    Windows are updated for every added/deleted sign.
    Solution:   Do not call update_debug_sign().  Only redraw when the line with
                the sign is visible.  (idea from neovim #9479)
```
Bug was initially reported in comments of https://github.com/vim/vim/issues/3802
but there were 2 different issues there, so I created this as a separate ticket.

### PR DESCRIPTION
**Pull request by jerdna-regeiz**
Integrates new sign api of Vim 8.1.
Took personal interest in having the functions of sign api in neovim (for vim-lsp).

#### vim-patch:8.1.0614: placing signs can be complicated

Problem:    Placing signs can be complicated.
Solution:   Add functions for defining and placing signs.  Introduce a group
            name to avoid different plugins using the same signs. (Yegappan
            Lakshmanan, closes vim/vim#3652)
https://github.com/vim/vim/commit/162b71479bd4dcdb3a2ef9198a1444f6f99e6843

#### vim-patch:8.1.0632: using sign group names is inefficient

Problem:    Using sign group names is inefficient.
Solution:   Store group names in a hash table and use a reference to them.
            Also remove unnecessary use of ":exe" from the tests.  (Yegappan
            Lakshmanan, closes vim/vim#3715)
https://github.com/vim/vim/commit/7a2d9892b7158edf8dc48e9bcaaae70a40787b37

#### vim-patch:8.1.0644: finding next sign ID is inefficient

Problem:    Finding next sign ID is inefficient.
Solution:   Add next_sign_id. (Yegappan Lakshmanan, closes vim/vim#3717)
https://github.com/vim/vim/commit/6436cd83f90a0efc326798792e49e8ff96a43dce

#### vim-patch:8.1.0658: deleting signs and completion for :sign is insufficient

Problem:    Deleting signs and completion for :sign is insufficient.
Solution:   Add deleting signs in a specified or any group from the current
            cursor location.  Add group and priority to sign command
            completion. Add tests for different sign unplace commands. Update
            help text.  Add tests for sign jump with group. Update help for
            sign jump. (Yegappan Lakshmanan, closes vim/vim#3731)
https://github.com/vim/vim/commit/7d83bf4f2b785b46d87c7bc376fc9d0a862af782

#### vim-patch:8.1.0660: sign_cleanup() may leak memory

Problem:    sign_cleanup() may leak memory.
Solution:   Free the group name before returning.  Add a few more tests.
            (Yegappan Lakshmanan)
https://github.com/vim/vim/commit/1ea88a3e12e5f6afc5353cd0d6e6d4f4a89ced94

#### vim-patch:8.1.0669: the ex_sign() function is too long

Problem:    The ex_sign() function is too long.
Solution:   Refactor the function.  Add a bit more testing. (Yegappan
            Lakshmanan, closes vim/vim#3745)
https://github.com/vim/vim/commit/a355652ea5b0c1633e8126ad9af2d970e05f4e1a

#### vim-patch:8.1.0673: functionality for signs is spread out over several files

Problem:    Functionality for signs is spread out over several files.
Solution:   Move most of the sign functionality into sign.c. (Yegappan
            Lakshmanan, closes vim/vim#3751)
https://github.com/vim/vim/commit/bbea47075cc4e7826e9f8c203e4272ba023ed7b0

#### vim-patch:8.1.0679: sign functions do not take buffer argument as documented

Problem:    Sign functions do not take buffer argument as documented.
Solution:   Use get_buf_tv(). (Yegappan Lakshmanan, closes vim/vim#3755)
https://github.com/vim/vim/commit/2cbc1a02cb72916dfdbd0d307512c7c3fb766edf

#### vim-patch:8.1.0697: ":sign place" requires the buffer argument

Problem:    ":sign place" requires the buffer argument.
Solution:   Make the argument optional.  Also update the help and clean up the
            sign test. (Yegappan Lakshmanan, closes vim/vim#3767)
https://github.com/vim/vim/commit/b328cca2548936c5f68fff683049a929882f5011

#### vim-patch:8.1.0701: sign message not translated and inconsistent spacing

Problem:    Sign message not translated and inconsistent spacing.
Solution:   Add _() for translation.  Add a space. (Ken Takata)  Also use
            MSG_BUF_LEN instead of BUFSIZ.
https://github.com/vim/vim/commit/d730c8e2974609034016ca66db09d2ef78784343

#### vim-patch:8.1.0702: ":sign place" only uses the current buffer

Problem:    ":sign place" only uses the current buffer.
Solution:   List signs for all buffers when there is no buffer argument.
            Fix error message for invalid buffer name in sign_place().
            (Yegappan Lakshmanan, closes vim/vim#3774)
https://github.com/vim/vim/commit/b589f95b38ddd779d7e696abb0ea011dc92ea903

#### vim-patch:8.1.0709: windows are updated for every added/deleted sign

Problem:    Windows are updated for every added/deleted sign.
Solution:   Do not call update_debug_sign().  Only redraw when the line with
            the sign is visible.  (idea from neovim vim/vim#9479)
https://github.com/vim/vim/commit/27a472c32ed5b5298bca50864570a4a71ec1d204

#### vim-patch:8.1.0717: there is no function for the ":sign jump" command

Problem:    There is no function for the ":sign jump" command.
Solution:   Add the sign_jump() function. (Yegappan Lakshmanan, closes vim/vim#3780)
https://github.com/vim/vim/commit/6b7b7190aa9e5c4f51bceaebf9275aa5097cfea1

#### vim-patch:8.1.0750: when the last sign is deleted the signcolumn may remain

Problem:    When the last sign is deleted the signcolumn may not be removed
            even though 'signcolumn' is "auto".
Solution:   When deleting the last sign redraw the buffer. (Dominique Pelle,
            closes vim/vim#3803, closes vim/vim#3804)
https://github.com/vim/vim/commit/8144acbec33b751788a7912e2d880c083c6cfe93

#### vim-patch:8.1.0767: when deleting lines at the bottom signs are misplaced

Problem:    When deleting lines at the bottom signs are misplaced.
Solution:   Properly update the line number of signs at the end of a buffer
            after a delete/undo operation. (Yegappan Lakshmanan, closes vim/vim#3798)
https://github.com/vim/vim/commit/c771bf901622064dc27421b04853e16b6914a295

#### vim-patch:8.1.0772: the sign_define_by_name() function is too long

Problem:    The sign_define_by_name() function is too long.
Solution:   Split it into smaller functions. (Yegappan Lakshmanan,
            closes vim/vim#3819)
https://github.com/vim/vim/commit/0314236aabcb2ca9d0b74074dadecf68d7c7ed5f

#### vim-patch:8.1.0039: cannot easily delete lines in another buffer

Problem:    Cannot easily delete lines in another buffer.
Solution:   Add deletebufline().
https://github.com/vim/vim/commit/d79a26219d7161e9211fd144f0e874aa5f6d251e

### CODE REVIEW

**Comment by jerdna-regeiz** [comment]
Was quite a pile of patches to get the sign functionality of vim-lsp working in neovim and it is the first time I tried to merge a vim.patch, so please can somebody do a careful review.
I'll quite happily apply changes.

vim-patch:8.1.0039 is not really required (and might better belong to an other group of patches), but was necessary to get the new "oldtests" working, which test some of the new functionality.

**Comment by justinmk** [comment]
> vim-patch:8.1.0039 is not really required (and might better belong to an other group of patches), but was necessary to get the new "oldtests" working, which test some of the new functionality.

That's totally fine, it's better to include previous patches than skip around.

**Comment by justinmk** [comment]
ASAN build found a problem:

```
    #0 0xf028ef in hash_lookup /home/travis/build/neovim/neovim/build/../src/nvim/hashtab.c:146:17
    #1 0x1827aa9 in sign_group_ref /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:94:8
    #2 0x1826ea3 in insert_sign /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:217:19
    #3 0x1814366 in insert_sign_by_lnum_prio /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:276:3
    #4 0x1813e22 in buf_addsign /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:341:5
    #5 0x181cbf1 in sign_place /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:930:5
    #6 0x9f36b3 in f_sign_place /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:15783:7
    #7 0x8f7737 in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6426:11
    #8 0x90c7d3 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6168:11
    #9 0x99af71 in eval7 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:4266:15
    #10 0x997513 in eval6 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3963:7
    #11 0x994e17 in eval5 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3814:7
    #12 0x99011f in eval4 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3551:7
    #13 0x98f5fc in eval3 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3469:7
    #14 0x98eacc in eval2 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3400:7
    #15 0x8efb52 in eval1 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3327:7
    #16 0x90c1ad in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6141:9
    #17 0x90535e in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2763:9
    #18 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
    #19 0xc56a97 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:594:20
    #20 0x92d3ac in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:22360:3
    #21 0x8f705b in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6411:11
    #22 0x90c7d3 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6168:11
    #23 0x90535e in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2763:9
    #24 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
    #25 0xc56a97 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:594:20
    #26 0x94ec0a in ex_execute /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:20482:7
    #27 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
    #28 0xc56a97 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:594:20
    #29 0x92d3ac in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:22360:3
    #30 0x8f705b in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6411:11
    #31 0x90c7d3 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6168:11
    #32 0x90535e in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2763:9
    #33 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
    #34 0xc56a97 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:594:20
    #35 0xc3c3c0 in do_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:3232:3
    #36 0xc381f4 in cmd_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2977:14
    #37 0xc389b0 in ex_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2958:3
    #38 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
    #39 0xc56a97 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:594:20
    #40 0xc5cc65 in do_cmdline_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:270:10
    #41 0xfec75b in exe_commands /home/travis/build/neovim/neovim/build/../src/nvim/main.c:1709:5
    #42 0xfd9ac1 in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:549:5
    #43 0x7fd5e647282f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #44 0x450ca8 in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x450ca8)

0x60200002cf1f is located 1 bytes to the right of 14-byte region [0x60200002cf10,0x60200002cf1e)
allocated by thread T0 here:
    #0 0x509328 in malloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x509328)
    #1 0x110d014 in try_malloc /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:70:15
    #2 0x110d1d4 in xmalloc /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:104:15
    #3 0x1827cb5 in sign_group_ref /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:98:28
    #4 0x1826ea3 in insert_sign /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:217:19
    #5 0x1814366 in insert_sign_by_lnum_prio /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:276:3
    #6 0x1813cf6 in buf_addsign /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:335:11
    #7 0x181cbf1 in sign_place /home/travis/build/neovim/neovim/build/../src/nvim/sign.c:930:5
    #8 0x9f36b3 in f_sign_place /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:15783:7
    #9 0x8f7737 in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6426:11
    #10 0x90c7d3 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6168:11
    #11 0x99af71 in eval7 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:4266:15
    #12 0x997513 in eval6 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3963:7
    #13 0x994e17 in eval5 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3814:7
    #14 0x99011f in eval4 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3551:7
    #15 0x98f5fc in eval3 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3469:7
    #16 0x98eacc in eval2 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3400:7
    #17 0x8efb52 in eval1 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3327:7
    #18 0x90c1ad in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6141:9
    #19 0x90535e in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2763:9
    #20 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
    #21 0xc56a97 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:594:20
    #22 0x92d3ac in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:22360:3
    #23 0x8f705b in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6411:11
    #24 0x90c7d3 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6168:11
    #25 0x90535e in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2763:9
    #26 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
    #27 0xc56a97 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:594:20
    #28 0x94ec0a in ex_execute /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:20482:7
    #29 0xc78310 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2244:5
```

File: src/nvim/buffer.c [hunk 1]
```diff
@@ -94,6 +94,8 @@ static char *e_auabort = N_("E855: Autocommands caused command to abort");
 // Number of times free_buffer() was called.
 static int buf_free_count = 0;
 
+void buf_delete_signs(buf_T *buf, char_u *group);
```
Comment (justinmk) [inline src/nvim/buffer.c]: can remove, forward-declarations are auto-generated.

File: src/nvim/eval.c [hunk 2]
```diff
@@ -7978,8 +8014,6 @@ static void f_dictwatcherdel(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   }
 
   if (argvars[0].v_type != VAR_DICT) {
-    emsgf(_(e_invarg2), "dict");
-    return;
```
Comment (justinmk) [inline src/nvim/eval.c]: why remove this? `f_dictwatcherdel` should not be related to signs stuff, Vim does not have `f_dictwatcherdel`.

File: src/nvim/buffer.c (hunk 1 above)
Comment (bew) [reply src/nvim/buffer.c]: I'm curious, how does that happen?

Comment (jerdna-regeiz) [reply src/nvim/buffer.c]: done

File: src/nvim/eval.c (hunk 2 above)
Comment (jerdna-regeiz) [reply src/nvim/eval.c]: I really don't know how this slipped in, sorry! Back again now.

File: src/nvim/buffer.c (hunk 1 above)
Comment (justinmk) [reply src/nvim/buffer.c]: @bew we have a build step that runs Lua code, see `src/nvim/generators`.

**Comment by justinmk** [comment]
@jerdna-regeiz  nicely done. We need to fix the ASAN issue https://github.com/neovim/neovim/pull/10059#issuecomment-495880271 , that could be a later vim-patch, maybe it can be found with some git-fu. Otherwise we'll need to fix it.

`make lint` has complaints, those can wait until the end, in a separate `lint` commit.

**Comment by jerdna-regeiz** [comment]
> @jerdna-regeiz nicely done. We need to fix the ASAN issue [#10059 (comment)](https://github.com/neovim/neovim/pull/10059#issuecomment-495880271) , that could be a later vim-patch, maybe it can be found with some git-fu. Otherwise we'll need to fix it.

Already looking at the code. How can I reproduce the ASAN output locally? Somehow valgrind won't run with the ASAN build (a github issue hints that there might be a issue with my kernel...).
The code looks the same in current vim master, with the differences of using xmalloc and the hash_lookup is different. So most likely there is no vim-patch.

> `make lint` has complaints, those can wait until the end, in a separate `lint` commit.
Yes, had that on the roadmap, but left it till the end (patches are matching better without...).

**Comment by jerdna-regeiz** [comment]
What shall be done about the failing sign memory allocation (old)test case?
`From test_signs.vim:
Found errors in Test_sign_memfailures():
Caught exception in Test_sign_memfailures(): Vim(execute):E121: Undefined variable: s:srcdir @ function RunTheTest[35]..Test_sign_memfailures[4]..GetAllocId, line 1
TEST FAILURE`

As far as I can see, dosn't neovim have this allocation id feature. Shall we remove the Test_sign_memfailures? Or can it be marked as known issue or similar? Or is there any other way to get this working?

**Comment by justinmk** [comment]
> What shall be done about the failing sign memory allocation (old)test case?

`:help E342` in Vim says:

```
...
is consuming your memory.  Be careful!  Vim is not completely prepared for an
out-of-memory situation.  First make sure that any changes are saved.  Then
try to solve the memory shortage.  To stay on the safe side, exit Vim and
start again.
```

Thus, all of the thousands of diligent memory checks were removed in Nvim. They're useless. So Nvim also doesn't have `test_alloc_fail`, and `Test_sign_memfailures` can be skipped by adding this line at the start:

    throw 'skipped: Nvim does not support test_alloc_fail()'

(Note: don't delete the test, we want the test files to be near-identical so that merges are easier.)

**Comment by justinmk** [comment]
> How can I reproduce the ASAN output locally? 

See https://github.com/neovim/neovim/blob/master/src/nvim/README.md#build-with-asan

**Comment by justinmk** [comment]
```
../src/nvim/sign.c:996:43: error: format "%ld" expects argument of type "long int", but argument 2 has type "long long int" [-Werror=format=]
     EMSGN(_("E157: Invalid sign ID: %ld"), sign_id);
```

Try `PRId64` instead:

     EMSGN(_("E157: Invalid sign ID: %" PRId64), sign_id);

File: runtime/doc/eval.txt [hunk 3]
```diff
@@ -7348,7 +7379,240 @@ shiftwidth()						*shiftwidth()*
...
+
+		The {name} can be a String or a Number.  The optional {dict}
+		argument specifies the sign attributes.  The following values
+		are supported:
+		    icon	full path to the bitmap file for the sign.
+		    linehl	highlight group used for the whole line the
+				sign is placed in.
+		    text	text that is displayed when there is no icon
+				or the GUI is not being used.
+		    texthl	highlight group used for the text item
```
Comment (bfredl) [inline runtime/doc/eval.txt]: By looking at the code, I think "numhl" is also supported already.

Comment (jerdna-regeiz) [reply runtime/doc/eval.txt]: True, have seen to that within the code but forgot about documentation.

**Comment by justinmk** [comment]
`:retab` will fix much of the lint complaints.

File: runtime/doc/eval.txt (hunk 3 above)
Comment (jerdna-regeiz) [reply runtime/doc/eval.txt]: done

**Comment by justinmk** [comment]
LGTM. If you plan to squash some of the fixup commits , just

**Comment by jerdna-regeiz** [comment]
> LGTM. If you plan to squash some of the fixup commits , I will wait.

Great, then I'll squash it now.

> Note: the quickbuild failure is bogus.

Yeah, thought so as travis is completely green.

**Comment by jerdna-regeiz** [comment]
Okay, all squashed and rebased to latest master.

**Comment by justinmk** [comment]
Why is https://github.com/neovim/neovim/pull/10059/commits/51563b70d7689321411b9393bb0c85d3f2cdecdd needed? We don't care about matching Vim behavior if it regresses a feature that we added. Which test fails? It can just be updated or ignored.

**Comment by jerdna-regeiz** [comment]
If I remember correctly it is Test_sign_group which fails, as multiple signs are added with the same type.
Neovim allowed only a single singe per column with the same type. Vim allowed it and does explicitly test for it.
I saw no reason why there shouldn't be multiple signs with the same type (especially that they now might have different groups) and it was the only way to get the test running - and no further tests failed within neovim tests.

**Comment by justinmk** [comment]
> I saw no reason why there shouldn't be multiple signs with the same type

I think it was to solve a question regarding which ones to display when there are multiple sign-columns. However, since it's not covered by tests, let's forget it for now.

Thanks for this!

**Comment by janlazo** [comment]
8.1.0039 isn't fully merged because it partly affects code for `setbufline` and `appendbufline`, included in 8.0.x patches. Any other patches here that affect `set_buffer_lines`?

**Comment by jerdna-regeiz** [comment]
No, only this one, as only deletion was required to get the test running.

**Comment by justinmk** [comment]
@jerdna-regeiz for future reference, we want to avoid partially merging patches, it makes it very difficult to reconcile the differences later.

File: src/nvim/sign.c [hunk 4]
```diff
@@ -0,0 +1,1746 @@
...
+/// Struct to hold the sign properties.
+typedef struct sign sign_T;
+
+struct sign
+{
+    sign_T      *sn_next;       // next sign in list
+    int         sn_typenr;      // type number of sign
+    char_u      *sn_name;       // name of sign
+    char_u      *sn_icon;       // name of pixmap
+# ifdef FEAT_SIGN_ICONS
```
Comment (janlazo) [inline src/nvim/sign.c]: Where is this defined? I can't find `#define FEAT_SIGN_ICONS` in the codebase so I assume this is dead code.

Comment (justinmk) [reply src/nvim/sign.c]: Yep, we should remove the guard

File: src/nvim/sign.c [hunk 5]
```diff
@@ -0,0 +1,1746 @@
...
+        sign_get_placed_in_buf(cbuf, 0, sign_id, sign_group, retlist);
+      }
+    }
+  }
+}
+
+# if defined(FEAT_SIGN_ICONS) || defined(PROTO)
+/// Allocate the icons.  Called when the GUI has started.  Allows defining
+/// signs before it starts.
+void sign_gui_started(void)
```
Comment (janlazo) [inline src/nvim/sign.c]: Where is this used? Dead code?

File: src/nvim/sign.c [hunk 6]
```diff
@@ -0,0 +1,1746 @@
...
+}
+
+/// Initialize the icon information for a new sign
+static void sign_define_init_icon(sign_T *sp, char_u *icon)
+{
+  xfree(sp->sn_icon);
+  sp->sn_icon = vim_strsave(icon);
+  backslash_halve(sp->sn_icon);
+# ifdef FEAT_SIGN_ICONS
+  if (gui.in_use) {
```
Comment (janlazo) [inline src/nvim/sign.c]: Is `FEAT_SIGN_ICONS` for Vim's GUIs only? Where is `gui` defined?

File: src/nvim/sign.c [hunk 7]
```diff
@@ -0,0 +1,1746 @@
...
+    out_flush();
```
Comment (janlazo) [inline src/nvim/sign.c]: ui_flush()

File: src/nvim/sign.c (hunk 5 above)
Comment (justinmk) [reply src/nvim/sign.c]: we should remove the guards.

**Pull request merged**